### PR TITLE
resolve: fix bug in visibility of shadowed use declarations

### PR DESCRIPTION
**Pull request by jseyfried**
This fixes a bug in which the visibility of a use declaration defining a name in one namespace (e.g. the value namespace) is overridden by a later use declaration defining the same name in the other namespace (e.g. the type namespace). For example,

``` rust
fn f() {}
pub mod bar {}

mod foo {
    use f; // This import should not be visible outside `foo`,
    pub use bar as f; // but it visible outside of `foo` because of this import.
}

fn main() { foo::f(); }
```

As the example demonstrates, this is a [breaking-change], but it looks unlikely to cause breakage in practice, and any breakage can be fixed by correcting visibility modifiers.

### CODE REVIEW

**Comment by sanxiyn** [comment]
cc #30159.

**Comment by jseyfried** [comment]
This also fixes #30159.

File: src/librustc_resolve/record_exports.rs [hunk 1]
```diff
@@ -130,13 +130,14 @@ impl<'a, 'b, 'tcx> ExportRecorder<'a, 'b, 'tcx> {
...
         for (name, import_resolution) in module_.import_resolutions.borrow().iter() {
-            if !import_resolution.is_public {
-                continue;
-            }
             let xs = [TypeNS, ValueNS];
             for &ns in &xs {
-                match import_resolution.target_for_namespace(ns) {
-                    Some(target) => {
+                if !import_resolution[ns].is_public {
+                    continue;
```
Comment (nrc) [inline src/librustc_resolve/record_exports.rs]: should this be a labelled continue? It looks like you should continue to the next import_resolution, not to the next namespace

File: src/librustc_resolve/resolve_imports.rs [hunk 2]
```diff
@@ -102,80 +102,61 @@ impl Target {
...
+    // is if the name is imported by exactly two `use` directives, one of which resolves to a
+    // value and the other of which resolves to a type.
+    pub outstanding_references: usize,
+    pub type_ns: NsImportResolution,
+    pub value_ns: NsImportResolution,
+}
+
+/// Records what we know about an imported name in a namespace (see `ImportResolution`).
+#[derive(Clone,Debug)]
+pub struct NsImportResolution {
```
Comment (nrc) [inline src/librustc_resolve/resolve_imports.rs]: rename to `ImportResolutionPerNamespace` or something? I don't like the `Ns` prefix.

**Comment by nrc** [comment]
LGTM, modulo the two comments

File: src/librustc_resolve/record_exports.rs (hunk 1 above)
Comment (jseyfried) [inline src/librustc_resolve/record_exports.rs]: No, if the import resolution is not public in one namespace it still could be public in the other namespace. For example, consider
```rust
mod foo {}
pub fn f() {}

use foo as bar;
pub use f as bar;
```
Here, there is a single `import_resolution` corresponding to `bar` that is private in the type namespace (i.e. `!import_resolution[TypeNS].is_public`) but public in the value namespace (i.e. `import_resolution[ValueNS].is_public`). If we skipped to the next import_resolution after seeing that the type namespace is private, we would not export `bar` in the value namespace.

File: src/librustc_resolve/resolve_imports.rs (hunk 2 above)
Comment (jseyfried) [inline src/librustc_resolve/resolve_imports.rs]: Done.

**Comment by jseyfried** [comment]
@nrc Thanks for the feedback. I addressed both comments.

**Comment by nrc** [comment]
Sounds good, thanks!

@bors: r+

**Pull request merged**